README.md
# AtticRadar

**FULL CREDIT GOES TO [netbymatt](https://github.com/netbymatt) for pretty much all of this project.**

He provided **FOUR** libraries:<br>[nexrad-level-2-data](https://github.com/netbymatt/nexrad-level-2-data)<br>[nexrad-level-2-plot](https://github.com/netbymatt/nexrad-level-2-plot)<br>[nexrad-level-3-data](https://github.com/netbymatt/nexrad-level-3-data)<br>[nexrad-level-3-plot](https://github.com/netbymatt/nexrad-level-3-plot)<br>that allowed for the development of this project. I only take credit for porting these Node apps to the browser with [Browserify](https://browserify.org).

<br>

**Pretty much all of the code that plots the data to the map using WebGL came from [QuadWeather's Radar Demo page](https://quadweather.com/radar-demo)**. Please go share him some love [on Twitter](https://twitter.com/quadweather).

His main radar page (which inspired this entire repository to be made) can be found here: [https://radar.quadweather.com](https://radar.quadweather.com)

<br><br>

In summary, the majority of this app was not made by me. I wanted to give the credit that was due to the people that made the frameworks of this app, because this website would not exist today if it weren't for the libraries / code snippets they provided. Here are their links again:

[netbymatt](https://github.com/netbymatt) (link to GitHub profile)<br>
[QuadWeather](https://twitter.com/quadweather) (link to Twitter profile)

# Setup
```
git clone https://github.com/SteepAtticStairs/AtticRadar.git
cd AtticRadar
npm install
npm run build
php -S 127.0.0.1:8080
```
then you can go to `localhost:8080` or `127.0.0.1:8080` to view the website.

(I used to use `python3 -m http.server 8080`, but it would frequently crash or freeze, which became extremely irritating. Although the php server functionality is really meant for php development, I have found that it works fine for this app, and it doesn't break like python's local server does.)

You can also run
```
npm run serve
```
to use `watchify` (a part of Browserify) to auto-bundle the project every time you make a change.

# Notes

**The website can be found live here:
<br>
https://steepatticstairs.github.io/AtticRadar/**
<br><br><br>
You can add some parameters to the URL to make the app easier to use / bookmark. You can add 
* `#station=ICAO`
* `#tideStation=TIDESTATIONID`

The first one is to set the app's initial station. This is the station the map would load radar products for / load METAR stations around, if there isn't a selected station. (The default will likely be either `KLWX` or `KMHX`, because those are the stations I use the most.)  The second one is to set the default tide station. Here are some examples:
<br>

Sets the app's initial station to `KHGX`:
```
https://steepatticstairs.github.io/AtticRadar/#station=KHGX
```
Sets the app's initial tide station to `8724698` (Loggerhead Key):
```
https://steepatticstairs.github.io/AtticRadar/#tideStation=8724698
```
Sets the app's initial station to `KHGX` AND sets the tide station to `8724698`:
```
https://steepatticstairs.github.io/AtticRadar/#station=KHGX&#tideStation=8724698
```
Notice how in the last example, the two parameters are separated by an ampersand (`&`) and still require a hash (`#`) before each parameter.


# History
* `Sep 10 2022` - Renamed this project from "NexradJS" to "AtticRadar"

CHANGELOG.md
**v4.1.0**\
`07/31/2022 10:05 AM EDT`<!-- 1659276319 --><br />[5b7ca39](https://github.com/SteepAtticStairs/AtticRadar/commit/5b7ca39738e667d346de961b7d3e9063e0f45c15)
* selecting a tilt now automatically loads that file onto the map

**v4.0.0**\
`07/30/2022 11:24 PM EDT`<!-- 1659237852 --><br />[4141b37](https://github.com/SteepAtticStairs/AtticRadar/commit/4141b370f28a51bc064939020963aff68fd56698)
* started working on redoing the app with bootstrap

---

**v3.15.0**\
`07/30/2022 09:16 PM EDT`<!-- 1659230179 --><br />[b9f1e3a](https://github.com/SteepAtticStairs/AtticRadar/commit/b9f1e3a4cda8e09e1cc4c0b8acde18c2b5eeb71d)
* started working on allowing storm tracks to be overlayed on L2 files
* cleanup

**v3.14.0**\
`07/30/2022 05:40 PM EDT`<!-- 1659217209 --><br />[ce4f1c7](https://github.com/SteepAtticStairs/AtticRadar/commit/ce4f1c7aefcc421e65a8736cad50ee9d30d959f6)
* added circles for each edge on a storm track line

**v3.13.0**\
`07/30/2022 05:26 PM EDT`<!-- 1659216371 --><br />[a4549b2](https://github.com/SteepAtticStairs/AtticRadar/commit/a4549b2a44d434c2bec93796f2d7712b4d683055)
* added point to the start of a line storm track

**v3.12.0**\
`07/30/2022 05:08 PM EDT`<!-- 1659215291 --><br />[56fd407](https://github.com/SteepAtticStairs/AtticRadar/commit/56fd4072019999212adb3eb0e11431d0a39f177b)
* started working on storm tracking information

**v3.11.0**\
`07/29/2022 10:46 PM EDT`<!-- 1659149216 --><br />[09b8735](https://github.com/SteepAtticStairs/AtticRadar/commit/09b873586ae6f32ba5fa4aafafc961f8e495e0bd)
* started working on adding L3 vertically integrated liquid product

**v3.10.0**\
`07/29/2022 04:21 PM EDT`<!-- 1659126106 --><br />[d0907a5](https://github.com/SteepAtticStairs/AtticRadar/commit/d0907a566a4785b3afaae775ee6ae031a0ffec62)
* added L3 base velocity product

**v3.9.0**\
`07/29/2022 02:27 PM EDT`<!-- 1659119245 --><br />[ea7725c](https://github.com/SteepAtticStairs/AtticRadar/commit/ea7725ccda167a79b5b382c35c3e7a2cd0b94c52)
* updated accuracy of L3 storm relative velocity plotting

**v3.8.0**\
`07/29/2022 02:25 PM EDT`<!-- 1659119105 --><br />[72bbcf5](https://github.com/SteepAtticStairs/AtticRadar/commit/72bbcf550934ddac9a8b579f8903258324055b67)
* added dynamic L3 product changing

**v3.7.1**\
`07/29/2022 02:11 PM EDT`<!-- 1659118301 --><br />[7eef8fb](https://github.com/SteepAtticStairs/AtticRadar/commit/7eef8fb92b29c8cb2a0059a8e4982fd79ec36ffa)
* updated L3 base reflectivity color palette

**v3.7.0**\
`07/29/2022 02:10 PM EDT`<!-- 1659118213 --><br />[d2fd9a0](https://github.com/SteepAtticStairs/AtticRadar/commit/d2fd9a01409cf24e84c455b512f24062d78d499d)
* added L3 base reflectivity product

**v3.6.0**\
`07/29/2022 01:18 PM EDT`<!-- 1659115130 --><br />[743edf9](https://github.com/SteepAtticStairs/AtticRadar/commit/743edf927f7f5190c3aaad841c8a7b0f30a389c7)
* added L2 TDWR support

**v3.5.0**\
`07/28/2022 06:48 PM EDT`<!-- 1659048528 --><br />[56dacbc](https://github.com/SteepAtticStairs/AtticRadar/commit/56dacbc61a5da6e2bdc7ace6e7d220a6000172c0)
* added L3 storm relative velocity product

**v3.4.1**\
`07/28/2022 05:47 PM EDT`<!-- 1659044824 --><br />[b095c43](https://github.com/SteepAtticStairs/AtticRadar/commit/b095c4307b15076c2677de71e15e378d5e4d32f0)
* updated L3 hydrometer classification color palete

**v3.4.0**\
`07/28/2022 12:41 PM EDT`<!-- 1659026462 --><br />[0f05f85](https://github.com/SteepAtticStairs/AtticRadar/commit/0f05f85602cd313591c93cec726a712a74a03221)
* added all L3 hydrometer classification product options

**v3.3.0**\
`07/27/2022 11:41 PM EDT`<!-- 1658979707 --><br />[907e097](https://github.com/SteepAtticStairs/AtticRadar/commit/907e097c29b24ca938f3b3129b51233d3559b7bc)
* added the ability to fetch and plot the latest tilt 1 L3 hydrometer classification file (N0H)

**v3.2.0**\
`07/27/2022 11:07 PM EDT`<!-- 1658977649 --><br />[30c89a9](https://github.com/SteepAtticStairs/AtticRadar/commit/30c89a9826526625394a46a7f80b633ee960ab94)
* fixed a bug with L3 azimuth generating

**v3.1.1**\
`07/27/2022 09:26 PM EDT`<!-- 1658971605 --><br />[457b6ee](https://github.com/SteepAtticStairs/AtticRadar/commit/457b6ee1b176c82cd94cae5f9f1477a39a026d58)
* added code to display a test L3 file (HHC) to the map

**v3.1.0**\
`07/27/2022 03:45 PM EDT`<!-- 1658951118 --><br />[06e6661](https://github.com/SteepAtticStairs/AtticRadar/commit/06e666164aaf630bef75a7530e038cc2cd9496aa)
* started to work on plotting level 3 (L3) files to the map

**v3.0.1**\
`07/27/2022 02:36 PM EDT`<!-- 1658946987 --><br />[36fd1ff](https://github.com/SteepAtticStairs/AtticRadar/commit/36fd1ffce446f9a8f338b9fe952ddc90deff8eaa)
* added nexrad-level-3-plot repo

**v3.0.0**\
`07/26/2022 10:51 PM EDT`<!-- 1658890277 --><br />[1c68e22](https://github.com/SteepAtticStairs/AtticRadar/commit/1c68e2251447f043fe29b68637005aa082ba88ab)
* added nexrad-level-3-data repo

---

**v2.10.0**\
`07/26/2022 09:20 PM EDT`<!-- 1658884801 --><br />[eecaffe](https://github.com/SteepAtticStairs/AtticRadar/commit/eecaffe83cf6d3bd0a3faa3e6f35a035810b0a07)
* added support for L2 file version 3

**v2.9.0**\
`07/26/2022 07:34 PM EDT`<!-- 1658878464 --><br />[e6bb9aa](https://github.com/SteepAtticStairs/AtticRadar/commit/e6bb9aa6abf1a56987a0e67b2200ea4c718d30e7)
* added L2 spectrum width (sw ) product

**v2.8.0**\
`07/26/2022 06:42 PM EDT`<!-- 1658875334 --><br />[641ccaf](https://github.com/SteepAtticStairs/AtticRadar/commit/641ccafbfcd2cdbf934cf5142933bcb385c8ced0)
* added L2 differential reflectivity (zdr) product

**v2.7.0**\
`07/26/2022 06:18 PM EDT`<!-- 1658873936 --><br />[a6516fe](https://github.com/SteepAtticStairs/AtticRadar/commit/a6516fe162e9375ebbd58bb8c65cb15b4ed29550)
* added L2 differential phase shift (phi) product

**v2.6.0**\
`07/26/2022 03:35 PM EDT`<!-- 1658864119 --><br />[fd9eda7](https://github.com/SteepAtticStairs/AtticRadar/commit/fd9eda7e755e6d4400515411a9fa0a7fd037f8f3)
* added L2 dynamic elevation displaying

**v2.5.0**\

app/loaders.js
const mapFuncs = require('./map/mapFunctions');
const ut = require('./utils');


// https://stackoverflow.com/a/64123890
// https://github.com/samundrak/fetch-progress
async function fetchWithProgress(url, callback) {
    const response = await fetch(url);
    let loaded = 0;

    const res = new Response(new ReadableStream({
        async start(controller) {
            const reader = response.body.getReader();
            for (;;) {
                const {done, value} = await reader.read();
                if (done) break;
                loaded += value.byteLength;
                ut.progressBarVal('label', ut.formatBytes(loaded));
                ut.progressBarVal('set', parseInt(ut.formatBytes(loaded)) / 10);
                controller.enqueue(value);
            }
            controller.close();
        },
    }));
    callback(await res.blob());
}

/**
* Function to load a radar plot onto the map from a URL.
*
* @param {any} url - The location to the file. This can be a relative path, or a URL
to a file hosted on a server.
* @param {any} level - What level file this is. This is a temporary parameter, it will be
removed when level detection is added. Use 2 for level 2, 3 for level 3, OR 22 for a level 2
file where you only want to load the first chunk of the file (reflectivity data) for a quicker
loading speed.
*/
function loadFileObject(url, level) {
    ut.progressBarVal('show');
    var radLevel;
    var wholeOrPart = 'whole';
    if (level == 2) {
        radLevel = 2;
    } if (level == 22) {
        radLevel = 2;
        wholeOrPart = 'part';
    } else if (level == 3) {
        radLevel = 3;
    }
    console.log('Fetch initialized - data requested');
    fetchWithProgress(url, function(resp) {
        console.log('File finished downloading');
        var response = resp;
        var blob;

        if (level != 2) {
            /*
            this block of code is an attempt to catch a level 3 file where the data doesn't
            start until 11 bytes out. This will detect if the first four bytes are "SDUS"
            (something like SDUS32) and if they are not, remove the first 11 bytes
            (the first 11 bytes are the bytes that should be removed to allow the parser to work)
            */
            // store the first eleven bytes for checking
            var fileEarlyBytes = ut.blobToString(response.slice(0, 11));
            // if the first four bytes are not "SDUS"
            if (fileEarlyBytes.slice(0, 4) != "SDUS") {
                // remove those pesky 11 bytes!
                blob = response.slice(11);
            } else {
                // the file is fine, proceed as normal
                blob = response;
            }
        } else {
            blob = response;
        }

        blob.lastModifiedDate = new Date();
        blob.name = url;
        // Create the event
        var event = new CustomEvent("loadFile", {
            "detail": [
                blob,
                radLevel,
                wholeOrPart
            ]
        });
        // Dispatch/Trigger/Fire the event
        document.dispatchEvent(event);
    })
}

/**
* Sub-function to get the latest Level 2 file for a station. This should not be used by itself,
* rather it should be called from the main function getLatestFile().
*
* @param {any} station - The four letter ICAO of the station. e.g. "KLWX" / "KMHX"
* @param {any} callback - The function to run after the retrieval. Use a single variable
in this function, this will be a string with the latest file's URL.
*/
function getLatestL2(station, callback) {
    //document.getElementById('spinnerParent').style.display = 'block';
    var curTime = new Date();
    var year = curTime.getUTCFullYear();
    var month = curTime.getUTCMonth() + 1;
    if (month.toString().length == 1) month = "0" + month.toString();
    var day = curTime.getUTCDate();
    if (day.toString().length == 1) day = "0" + day.toString();
    var stationToGet = station.toUpperCase().replace(/ /g, '')
    var fullURL = "https://noaa-nexrad-level2.s3.amazonaws.com/?list-type=2&delimiter=%2F&prefix=" + year + "%2F" + month + "%2F" + day + "%2F" + stationToGet + "%2F"
    //console.log(fullURL)
    var baseURL = 'https://noaa-nexrad-level2.s3.amazonaws.com';
    //https://noaa-nexrad-level2.s3.amazonaws.com/2022/08/09/KATX/KATX20220809_004942_V06
    $.get(ut.phpProxy + fullURL, function (data) {
        var dataToWorkWith = JSON.stringify(ut.xmlToJson(data)).replace(/#/g, 'HASH')
        dataToWorkWith = JSON.parse(dataToWorkWith)
        //console.log(dataToWorkWith)
        var filenameKey = dataToWorkWith.ListBucketResult.Contents
        var latestFileName = filenameKey[filenameKey.length - 1].Key.HASHtext.slice(16);
        if (latestFileName.includes('MDM')) {
            latestFileName = filenameKey[filenameKey.length - 2].Key.HASHtext.slice(16);
        }

        var finishedURL = `${baseURL}/${year}/${month}/${day}/${station}/${latestFileName}`;
        callback(finishedURL);
    })
}

/**
* Sub-function to get the latest Level 3 file for a station. This should not be used by itself,
* rather it should be called from the main function getLatestFile().
*
* @param {any} station - The four letter ICAO of the station. e.g. "KLWX" / "KMHX"
* @param {any} product - Three letter abbreviation of the Level 3 product being retrieved. e.g. "NST", "N0B", "N0G"
* @param {any} callback - The function to run after the retrieval. Use a single variable
in this function, this will be a string with the latest file's URL.
*/
function getLatestL3(station, product, callback) {
    // if (!(product.length > 3)) {
    //     //document.getElementById('spinnerParent').style.display = 'block';
    //     var curTime = new Date();
    //     var year = curTime.getUTCFullYear();
    //     var month = curTime.getUTCMonth() + 1;
    //     if (month.toString().length == 1) month = "0" + month.toString();
    //     var day = curTime.getUTCDate();
    //     if (day.toString().length == 1) day = "0" + day.toString();
    //     var stationToGet = station.toUpperCase().replace(/ /g, '')
    //     var urlBase = "https://unidata-nexrad-level3.s3.amazonaws.com/";
    //     var filenamePrefix = `${station}_${product}_${year}_${month}_${day}`;
    //     // var urlPrefInfo = '?list-type=2&delimiter=/%2F&prefix=';
    //     var urlPrefInfo = '?prefix=';
    //     var fullURL = `${urlBase}${urlPrefInfo}${filenamePrefix}`
    //     console.log(fullURL)
    //     $.get(ut.phpProxy + fullURL, function (data) {
    //         var dataToWorkWith = JSON.stringify(ut.xmlToJson(data)).replace(/#/g, 'HASH')
    //         dataToWorkWith = JSON.parse(dataToWorkWith)
    //         //console.log(dataToWorkWith)
    //         var contentsBase = dataToWorkWith.ListBucketResult.Contents;
    //         var filenameKey = contentsBase[contentsBase.length - 1].Key.HASHtext;

    //         var finishedURL = `${urlBase}${filenameKey}`;
    //         callback(finishedURL);
    //     })
    // } else {
    //     var fileUrl = `https://tgftp.nws.noaa.gov/SL.us008001/DF.of/DC.radar/DS.${product}/SI.${$('#stationInp').val().toLowerCase()}/sn.last`
    //     callback(fileUrl);
    // }

    /*
    * Below is all unused code to retrieve the latest file from a different data source.
    */
    var curTime = new Date();
    var year = curTime.getUTCFullYear();
    var month = curTime.getUTCMonth() + 1;
    if (month.toString().length == 1) month = "0" + month.toString();
    var day = curTime.getUTCDate();
    if (day.toString().length == 1) day = "0" + day.toString();
    var yyyymmdd = `${year}${month}${day}`
    var l3FileURL = `https://unidata3.ssec.wisc.edu/native/radar/level3/nexrad/${product}/${station}/${yyyymmdd}/`;
    console.log(l3FileURL);
    $.get(ut.phpProxy + l3FileURL, function(data) {
        var div = document.createElement('div')
        div.innerHTML = data;
        var jsonWithFileList = JSON.parse(ut.html2json(div));
        var fileListLength = jsonWithFileList.children[2].children.length;
        var filenameKey = jsonWithFileList.children[2].children[fileListLength - 2].attributes[0][1];

        var finishedURL = `${l3FileURL}${filenameKey}`;
        callback(finishedURL);
    })
}

/**
* Fetches the filename of the latest file for a given station.
*
* @param {any} station - The four letter ICAO of the station. e.g. "KLWX" / "KMHX"
* @param {any} levelProduct - This parameter is an object that specifies if the user is
retrieving a level 2 file (use 2 for this parameter), or a level 3 file (in that case,
use an array with two values, specifying both the level (3), and the product you want to retrieve. 
e.g. [3, "NST"])
* @param {any} callback - The function to run after the retrieval. Use a single variable
in this function, this will be a string with the latest file's URL.
*/
function getLatestFile(station, levelProduct, callback) {
    ut.progressBarVal('show');
    // obviously, the user wants a level 2 file
    if (levelProduct == 2) {
        getLatestL2(station, function(url) {
            callback(url);
        })
    } else {
        // assume this is a level 3 file - and fail if it is not an array
        if (!Array.isArray(levelProduct)) {
            throw new Error('You must provide an array for a level 3 product/level parameter.');
        }
        const product = levelProduct[1];
        /* we need to slice(1) here (remove the first letter) because the level 3 source we
        * are using only accepts a three character ICAO, e.g. "MHX" / "LWX" */
        getLatestL3(station.slice(1), product, function(url) {
            callback(url);
        })
    }
};

module.exports = {
    loadFileObject,
    getLatestFile
}